src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,250 @@
+  }
+  if (!!users) {
+    if (!principals.users) {
+      principals.users = [];
+    }
+    principals.users = Array.from(new Set([...principals.users, ...users]));
+  }
+  if (!!groups) {
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deleteFromPrincipals = (principals?: Principals, users?: string[], groups?: string[]) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const checkPermission = (currentPrincipals: Principals | undefined, principals: Principals) => {
+  return (
+    (currentPrincipals?.users &&
+      principals?.users &&
+      checkPermissionForSinglePrincipalType(currentPrincipals.users, principals.users)) ||
+    (currentPrincipals?.groups &&
+      principals?.groups &&
+      checkPermissionForSinglePrincipalType(currentPrincipals.groups, principals.groups))
+  );
+};
+
+const checkPermissionForSinglePrincipalType = (
+  currentPrincipalArray: string[],
+  principalArray: string[]
+) => {
+  return (
+    currentPrincipalArray &&
+    principalArray &&
+    (currentPrincipalArray.includes('*') ||
+      principalArray.some((item) => currentPrincipalArray.includes(item)))
+  );
+};
+
+export class ACL {
+  private permissions?: Permissions;
+  constructor(initialPermissions?: Permissions) {
+    this.permissions = initialPermissions || {};
+  }
+
+  // parse the permissions object to check whether the specific principal has the specific permission types or not
+  public hasPermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || permissionTypes.length === 0 || !this.permissions || !principals) {
+      return false;
+    }
+
+    const currentPermissions = this.permissions;
+    return permissionTypes.some((permissionType) =>
+      checkPermission(currentPermissions[permissionType], principals)
+    );
+  }
+
+  // permissions object build function, add principal with specific permission to the object
+  public addPermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || !principals) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
+      this.permissions[permissionType] = addToPrincipals(
+        this.permissions[permissionType],
+        principals.users,
+        principals.groups
+      );
+    }
+
+    return this;
+  }
+
+  // permissions object build function, remove specific permission of specific principal from the object
+  public removePermission(permissionTypes: string[], principals: Principals) {
```

Comment:
one question about `removePermission`:

From the documentation, it seems like write permission includes the read permission. 
For example, if a user have both read and write permission to the saved object, and we call `removePermission` to remove the read permission. In this case, is the write permission still valid?

Or should we set a rule that we can not remove read permission while there is still a valid write permission? Or if the rules have been set, then can we add some comments to explain it?

Reply:
For acl permission, it's either read or write, user won't have both permissions.  It's the workspace saved object permission check function that has this logic. For example,  if a user need read permission on a saved object, saved object permission will call ACL hasPermission function to check whether user has either read or write permission on that saved object. removePermission just remove the specified permission. We'll add comment to make it clear.

Reply:
For acl permission, it's either read or write, user won't have both permissions.  It's the workspace saved object permission check function that has this logic. For example,  if a user need read permission on a saved object, saved object permission will call ACL hasPermission function to check whether user has either read or write permission on that saved object. removePermission just remove the specified permission.

Reply:
@xluo-aws would you be able to add some inline comments like the structure defined here: https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5075#discussion_r1339606507

Reply:
Sure